File: src/services/ArchetypeService.ts
import axios from "axios";
import {RootPath} from "./RootPath";
import Archetype from "../models/actor/player/Archetype";

export default class ArchetypeService {
    static async createArchetype(name: string): Promise<Archetype> {
        return await (await axios.post(RootPath.Archetype + name)).data;
    }

    static async getArchetype(name: string): Promise<Archetype> {
        return await (await axios.get(RootPath.Archetype + name)).data;
    }

    static async updateArchetype(name: string, archetype: Archetype): Promise<Archetype> {
        return await (await axios.put(RootPath.Archetype + name, archetype)).data;
    }

    static async getArchetypes(): Promise<Archetype[]> {
        return await (await axios.get(RootPath.Archetype)).data;
    }
}

File: server/src/main/java/com/github/genraven/genesys/service/ArchetypeService.java
package com.github.genraven.genesys.service;

import com.github.genraven.genesys.domain.actor.player.Archetype;
import com.github.genraven.genesys.repository.ArchetypeRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
public class ArchetypeService {
    private final ArchetypeRepository archetypeRepository;

    @Autowired
    public ArchetypeService(final ArchetypeRepository archetypeRepository) {
        this.archetypeRepository = archetypeRepository;
    }

    public Flux<Archetype> getAllArchetypes() {
        return archetypeRepository.findAll();
    }

    public Mono<Archetype> getArchetype(final String name) {
        return archetypeRepository.findById(name);
    }

    public Mono<Archetype> createArchetype(final String name) {
        return archetypeRepository.save(new Archetype(name));
    }

    public Mono<Archetype> updateArchetype(final String name, final Archetype archetype) {
        return getArchetype(name).map(arch -> {
            arch.setDescription(archetype.getDescription());
            arch.setBrawn(archetype.getBrawn());
            arch.setAgility(archetype.getAgility());
            arch.setIntellect(archetype.getIntellect());
            arch.setCunning(archetype.getCunning());
            arch.setWillpower(archetype.getWillpower());
            arch.setPresence(archetype.getPresence());
            arch.setWounds(archetype.getWounds());
            arch.setStrain(archetype.getStrain());
            arch.setExperience(archetype.getExperience());
            arch.setSkill(archetype.getSkill());
            arch.setAbilities(archetype.getAbilities());
            return arch;
        }).flatMap(archetypeRepository::save);
    }
}
